### ISSUE TITLE
Use constant time by default for BIGNUMs

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#6078]
The BIGNUM library defaults to a non-constant time implementation of various operations. Where we want to force a constant time implementation we have to remember to set the BN_FLG_CONSTTIME flag. Experience has shown that we often forget to do this. For example see:

- 54f007a (CVE-2018-0737)
- 8db7946
- e913d11
- 6364475
- 6364475
- 3de81a5
- 47ae05b
- 033dc8f
- 3999446 (CVE-2016-2178)

This is not intended to be an exhaustive list - there are probably others cases that I haven't listed.

The difficulty of changing the default is that this will have a performance impact for cases where constant time is not a requirement. It is unclear to me how many of those cases there are and whether we should be worried about it or not.

PR #5969 made one attempt at fixing this. Feedback on that suggested a better approach may be to use separate APIs for constant-time and non-constant time operation:

@briansmith wrote:
> See https://github.com/HACS-workshop/spectre-mitigations/blob/master/crypto_guidelines.md#1-do-not-conditionally-choose-between-constant-and-non-constant-time. I'm biased, but I think BoringSSL's solution, which was to use separate APIs for constant-time and non-constant-time implementations, is better, in general.

**Comment by triska** [comment]
In my view, the following quote from Daniel Bernstein's [*Some thoughts on security after ten years of qmail&nbsp;1.0*](http://cr.yp.to/qmail/qmailsec-20071101.pdf) expresses the situation perfectly:

<blockquote>
I don’t like waiting for my computer. I really don’t like
waiting for someone else’s computer. [...] But I find security
much more important than speed. We <i>need</i> invulnerable
software systems, and we need them today, <b>even if they are
ten times slower than our current systems.</b>
</blockquote>

Speaking purely as a user of OpenSSL: If the approach you implement provides stronger and more valuable security guarantees, I will use it immediately, whether it is twice, thrice, or ten times slower than the current approach. So, thank you for taking this issue seriously, and thumbs up!

For comparison, here are some **GMP** low-level interface functions for cryptography:

[**8.1 Low-level functions for cryptography**](https://gmplib.org/manual/Low_002dlevel-Functions.html#index-Low_002dlevel-functions-for-cryptography)

These functions are resilient to side-channel attacks, and I hope that their design, implementation and trade-offs may provide some inspiration.

**Comment by mattcaswell** [comment]
See also #6640.

**Comment by nhorman** [comment]
Is this a duplicate of openssl/openssl#6640

**Comment by davidben** [comment]
No. #6640 is about how OpenSSL's `BIGNUM`s completely fail to be constant-time due to the in-memory representation it chose, causing numerous constant-time related security issues. The fix is to rework either `BIGNUM` or your cryptography primitives' internals to avoid this.

This issue is about how the `BN_FLG_CONSTTIME` flag is extremely error-prone and caused numerous _other_ constant-time related security issues. The fix proposed in this bug is to change the "default", but I actually think that's the wrong lesson. In BoringSSL, we removed the flag altogether. Rather, I think the right observation is that windowed exponentiation with fixed and sliding windows are _separate_ operations and should be separate functions altogether. If you're in a context where the exponent is secret, used fixed windows with a constant-time lookup in your table of precomputed multiples. If the exponent is public, used sliding windows to improve performance. If the exponent is not only public, but a constant, you may be even better off specializing to the exponent and hardcoding an optimal addition chain.
